orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,257 @@ common::Status HandleSharedInitializer(Graph& graph,
+      assert(found_stage != op_to_stage.end());
+      int consumer_stage = found_stage->second;
+      // TODO: test case in which a tensor is produced by a forward op, stashed
+      // and sent to the previous stage by a backward op.
+      // For now, I'm assuming that if a tensor is produced by a forward op and
+      // consumed by a backward op, then producer and consumer are both in the
+      // same device. This will not always be the case though. 
+      if (!IsBackward(*producer_node) && IsBackward(*consumer)) {
+        // They must be on the same device. 
+        ORT_ENFORCE(producer_stage == consumer_stage,
+          "Forward producer and backward consumer of a tensor must be in the same device.");
+      }
+
+      // It is impossible to have a backward operator producing a tensor
+      // consumed by a forward operator. So, at this point, either both producer
+      // and consumer are forward or both are backward. Either way, we want
+      // to know where are the last consumers of a tensor.
+      if (is_forward(producer_stage, consumer_stage)) {
+        last_consumer_stage_forward = std::max(last_consumer_stage_forward, consumer_stage);
+      } 
+      ORT_ENFORCE(!is_backward(producer_stage, consumer_stage),
+                  "Not supported yet: ", producer_stage, "->", consumer_stage);
+      // TODO(jufranc): we will need something like the following, where
+      // else if (is_backward(producer_stage, consumer_stage)) {
+      //  last_consumer_stage_backward is init to INT_MAX, for training graphs.
+      //  last_consumer_stage_backward = std::min(last_consumer_stage_backward, consumer_stage);
+      // }
+
+      // Find which tensors need to be sent to the next stage (if it is a forward
+      // message).
+      if (producer_stage + 1 == consumer_stage) {
+        forward_messages.at(producer_stage).insert(node_arg);
+      } 
+      // TODO(jufranc): find which tensors need to be sent to the previous stage
+      // (if it is a backward message). Something like:
+      // else if (producer_stage - 1 == consumer_stage) { 
+      //   backward_messages.at(producer_stage).insert(node_arg);
+      // }
+    }
+
+    // Create all the replicas for this tensor now. We also keep track of which
+    // tensors need to be forwarded, and their producer-consumer stage range.
+    // The replica of the tensor in the producer stage, is the tensor itself.
+    replicas[producer_stage] = const_cast<NodeArg *>(node_arg);
+    if (is_forward(producer_stage, last_consumer_stage_forward)) {
+      for (int r = producer_stage + 1; r <= last_consumer_stage_forward; ++r) {
+        create_tensor_replica(node_arg, r);
+      }
+      if (last_consumer_stage_forward - producer_stage > 1) {
+        forwarded_tensors.push_back({node_arg, 
+                                    {producer_stage, last_consumer_stage_forward}});
+      }
+    }
+    // TODO(jufranc): take care of is_backward case.
+  }
```

Comment:
Let's throw if it's backward case for now?

Reply:
agreed.